jxl/src/api/output.rs

```diff
@@ -5,95 +5,169 @@
 
 #![allow(unsafe_code)]
 
-use std::{mem::MaybeUninit, ops::DerefMut};
+use core::slice;
+use std::{marker::PhantomData, mem::MaybeUninit, ops::Range};
 
-/// Trait to represent image output buffers.
-/// This trait is implemented for
-/// - [u8], which will be interpreted as contiguous rows of data, starting from row 0 and
-///   followed by row 1, 2, ... without any gaps)
-/// - [&mut [u8]], which will be interpreted as buffers for distinct rows (starting from
-///   the buffer for row 0, then row 1, ...).
-pub trait JxlOutputBuffer<'a> {
-    /// Returns buffers for each row of the image. Buffers might be uninitialized, but the caller
-    /// guarantees that data that *is* initialized in the buffers is not overwritten with
-    /// uninitialized data.
-    ///
-    /// # Safety
-    /// The returned buffers must not be populated with uninit data.
-    unsafe fn get_row_buffers(
-        &'a mut self,
-        shape: (usize, usize),
-        bytes_per_pixel: usize,
-    ) -> impl DerefMut<Target = [&'a mut [MaybeUninit<u8>]]>;
+use num_traits::ToBytes;
+
+use crate::image::Image;
+
+pub struct JxlOutputBuffer<'a> {
+    // Safety invariants:
+    //  - uninit data is never written to `buf`.
+    //  - `buf` is valid for writes for all bytes in the range
+    //    `buf[i*bytes_between_rows..i*bytes_between_rows+bytes_per_row]` for all values of `i`
+    //    from `0` to `num_rows-1`.
+    //  - `self` has exclusive (write) access to the bytes in these ranges.
+    //  - all the bytes in those ranges (and in between) are part of the same allocated object.
+    //  - `num_rows > 0`, `bytes_per_row > 0`, `bytes_per_row <= bytes_between_rows`.
+    //  - The computation `E = bytes_between_rows * (num_rows-1) + bytes_per_row` does not
+    //    overflow and has a result that is at most `isize::MAX`.
+    buf: *mut MaybeUninit<u8>,
```

Comment:
wouldn't using ```&mut [V]```, where ```V: AsMut<[u8]>``` as in the render-pipeline be better? then it is up to the caller to initialize it or not

Reply:
`[MaybeUninit<u8>]` does not implement `AsMut<[u8]>` (or the other way round)